File: paint.py
# linear time and space complexity
def minCost(costs):
    if not costs:
        return 0

    n = len(costs)
    
    # Initialize a 2D DP array to store the minimum cost at each house for each color.
    dp = [[0] * 3 for _ in range(n)]
    
    # Initialize the DP array with the costs for the first house.
    dp[0] = costs[0]
    
    # Iterate through the remaining houses, updating the DP array.
    for i in range(1, n):
        dp[i][0] = costs[i][0] + min(dp[i-1][1], dp[i-1][2])
        dp[i][1] = costs[i][1] + min(dp[i-1][0], dp[i-1][2])
        dp[i][2] = costs[i][2] + min(dp[i-1][0], dp[i-1][1])
    
 
    return min(dp[-1])


costs = [[17,2,17],[16,16,5],[14,3,19]]
minimum_cost = minCost(costs)
print("Minimum cost to paint all houses:", minimum_cost)

File: paint.java
// Time Complexity :O(n*n) 
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :no
public class Solution {
    /**
     * @param costs: n x 3 cost matrix
     * @return: An integer, the minimum cost to paint all houses
     */
    public int minCost(int[][] costs) {
        // write your code here
        if(costs.length==0)
            return 0;
        return helper(costs);
    }
    private int helper(int[][]matrix){
        int n=matrix.length-1;
        for(int i=n-1;i>=0;i--){
            for(int j=0;j<matrix[i].length;j++){
                int add=matrix[i+1][j];
                if(j==0) //if red
                    add=Math.min(matrix[i+1][j+1],matrix[i+1][j+2]);
                else if(j==matrix[i].length-1) //if green
                    add=Math.min(matrix[i+1][j-2],matrix[i+1][j-1]);
                else //if blue
                    add=Math.min(matrix[i+1][j+1],matrix[i+1][j-1]);
                matrix[i][j]+=add;
                    
            }
        }
        System.out.println(Arrays.toString(matrix[0]));
        int min=matrix[0][0];
        for(int i=1;i<matrix[0].length;i++)
            if(min>matrix[0][i])
                min=matrix[0][i];
        return min;
    }
}